rfcs/20201113-tf-core.md

```diff
@@ -0,0 +1,174 @@
+
+Minimal and stable:  The set of APIs should be small. This helps reduce the burden of maintenance, and also helps keep the APIs more stable.  Another way to put this principle is orthogonality, which means that two instructions are orthogonal if they cannot be used to accomplish the same task. APIs which lack orthogonality naturally lead to cognitive burden for users, not the least of which is the dreaded feeling of confusion: “there’s more than one way to do things”.
+
+Backwards compatibility: As a result of this effort, no Python user code will need to change. This is basically a refactoring and implementation layering exercise within Tensorflow that would improve the overall quality and maintainability of the product.  Additionally, we will migrate some of the implementation from Python to C++. A set of new C++/C APIs will be created without affecting the existing Python APIs. 
+
+### Keys to get there
+
+Understand users’ needs
+
+Core team will constantly talk with users, get users involved for all the phases, starting from requirements, design phase, until the launch phase and afterward. We care about all core users and their feedback. Meanwhile the core team will set up a few high touch teams to work together. 
+
+Extensible and composable
+
+In our principle, we mentioned that the set of APIs needs to be minimal, providing everything directly is not the goal, and also not sustainable. Meanwhile, this minimal set of APIs should also be sufficient for users to build their library without hacking into core APIs, or requesting to change/add more APIs. So, it sounds contradictory to have both principles at the same time with each other. 
+
+The key here is to provide an extensible and composable mechanism, together with the set of core APIs.
+
+Performance
+
+TF Core will be extremely lightweight and performant. We’ll try to do as much computation as possible in C++ with a thin layer on top. For example, Python APIs will be one thin layer on top. This also involves building a clean C API layer where state of the art runtimes like TFRT and Graph Compilers like MLIR can plug in and improve performance. 
+
+
+### Guidelines on evaluating the core APIs
+
+* Is this API serving any use cases?
+
+When we add/remove APIs, the first thing we should ask is ‘what is the use case for this APIs’. This API should serve users’ purpose, not developers’ purpose. 
+
+* Is this API similar or redundant with existing ones? 
+
+Corresponding to the principle ‘Primitive/Essential/Orthogonality’: If one API can be written by using existing a few APIs together with mechanics core provides, then the API is not essential. Instead of including that particular APIs, the primitive APIs that API is composed of should be included in the core.
+
+* Is this API easy to use?
+  Usability - representation
```

Comment:
Some formatting (especially line breaks) are broken in this doc when rendered. For example, you need a blank line here to break the line.